File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        ansLen = 0
        l_final, r_final = 0,0
        
        for i in range(len(s)):
            
            #odd len
            l, r = i, i
            while l>=0 and r<len(s) and s[l] == s[r]:
                if r-l + 1 > ansLen:
                    ansLen = r-l + 1
                    l_final, r_final = l, r
                    #because it stores a copy everytime and thus O(n^3)
                    #res = s[l:r+1]
                l -= 1
                r += 1
                
            
            
            #even len
            l, r = i, i+1
            while l>=0 and r<len(s) and s[l] == s[r]:
                if r-l + 1 > ansLen:
                    ansLen = r-l + 1
                    l_final, r_final = l, r
                    #res = s[l:r+1]
                l -= 1
                r += 1
            
        return s[l_final:r_final+1]
            

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        
        string ans;
        int maxLen = 0, ans_l, ans_r;
        
        for(int i=0; i<n; i++){
            
            //odd
            int left = i, right = i;
            while(left >=0 and right < n and s[left] == s[right]){
                int currLen = right - left + 1;
                if(currLen > maxLen){
                    maxLen = currLen;
                    ans_l = left, ans_r = right;
                }
                left--;
                right++;
            }
            
            //even
            left = i, right = i+1;
            while(left >=0 and right < n and s[left] == s[right]){
                int currLen = right - left + 1;
                if(currLen > maxLen){
                    maxLen = currLen;
                    ans_l = left, ans_r = right;
                }
                left--;
                right++;
            }
        }
        return s.substr(ans_l, maxLen);
    }
};